lib/fake_idp/application.rb

```diff
@@ -26,60 +24,44 @@ def configuration
-
-    def configure_cert_and_keys
-      self.x509_certificate = idp_certificate
-      self.secret_key = configuration.idp_secret_key
-      self.algorithm = configuration.algorithm
-    end
-
-    def certificate
-      Base64.encode64(configuration.certificate).delete("\n")
-    end
-
-    def idp_certificate
-      Base64.encode64(configuration.idp_certificate).delete("\n")
-    end
-
-    def user_attrs
-      signed_in_user_attrs.merge(configuration.additional_attributes)
+    def build_xml_saml_response
+      FakeIdp::SamlResponse.new(
+        name_id: configuration.name_id,
+        issuer_uri: configuration.issuer,
+        saml_acs_url: @saml_acs_url, # Defined in #decode_SAMLRequest in ruby-saml-idp gem
+        saml_request_id: @saml_request_id, # Defined in #decode_SAMLRequest in ruby-saml-idp gem
+        user_attributes: user_attributes,
+        algorithm_name: configuration.algorithm,
+        certificate: configuration.idp_certificate,
+        secret_key: configuration.idp_secret_key,
+        encryption_enabled: configuration.encryption_enabled,
+      ).build
     end
 
-    def signed_in_user_attrs
+    def user_attributes
       {
         uuid: configuration.sso_uid,
         username: configuration.username,
         first_name: configuration.first_name,
         last_name: configuration.last_name,
-        email: configuration.email
-      }
-    end
-
-    def name_id
-      configuration.name_id
+        email: configuration.email,
+      }.merge(configuration.additional_attributes)
     end
 
-    def mock_saml_request
-      current_gem_dir = File.dirname(__FILE__)
-      sample_file_name = "#{current_gem_dir}/sample_init_request.txt"
-      File.write(sample_file_name, params[:SAMLRequest]) if params[:SAMLRequest]
-      File.read(sample_file_name).strip
+    # An AuthRequest is required by the ruby-saml-idp gem to begin the process of returning
+    # a SAMLResponse. We will likely remove the ruby-saml-idp dependency in a future update
+    def generate_saml_request
+      auth_request = OneLogin::RubySaml::Authrequest.new
+      auth_url = auth_request.create(saml_settings)
+      CGI.unescape(auth_url.split("=").last)
     end
 
-    def attributes_statement(attributes)
-      attributes_xml = attributes_xml(attributes).join
-
-      %[<saml:AttributeStatement>#{attributes_xml}</saml:AttributeStatement>]
-    end
-
-    def attributes_xml(attributes)
-      attributes.map do |name, value|
-        attribute_value = %[<saml:AttributeValue>#{value}</saml:AttributeValue>]
-
-        %[<saml:Attribute Name="#{name}">#{attribute_value}</saml:Attribute>]
+    def saml_settings
+      OneLogin::RubySaml::Settings.new.tap do |setting|
+        setting.assertion_consumer_service_url = configuration.callback_url
+        setting.issuer = configuration.issuer
+        setting.idp_sso_target_url = configuration.idp_sso_target_url
+        setting.name_identifier_format = FakeIdp::SamlResponse::EMAIL_ADDRESS_FORMAT
```

Comment:
Does this mean that we're enforcing users of the library to always use this format? Should it be configurable, but with a default?

Reply:
Yes, we would be forcing this for the AuthRequest but not for the SAML response which is what we really care about for the sake of using the gem for testing. The AuthRequest here could theoretically be completely skipped or stubbed since in the wild it's used for the authentication of a user at the SP-initiated end of an SSO workflow. However, it does not have to impact the SAML response attributes. I kept it here because it could be a good starting point for expanding the gem to support end to end testing of an SP-initiated SSO workflow and is a bit more meaningful than what we had before.

I don't see any value making it configurable right now since it doesn't affect the generated assertion but let me know if you think otherwise.

Reply:
I'm ok with it as-is; I wanted clarification on the implication. Thanks!